File: product_except_self.rb
# giai thuat: tich cac phan tu ben trai * tich cac phan tu ben phai
def product_except_self(nums)
  results = Array.new(nums.size, 1)
  left, right = 1, 1
  (0...nums.size).each do |i|
    results[i] = left
    left *= nums[i]
  end
  (nums.size - 1).downto(0) do |i|
    results[i] *= right
    right *= nums[i]
  end
  results
end

puts product_except_self([1, 2, 3, 4]).inspect # => [24, 12, 8, 6]
puts product_except_self([1, 0, 3, 4]).inspect  # [0, 12, 4, 3]

File: Python/product_except_self.py
def productExceptSelf(nums):
    n = len(nums)
    results = [1] * n
    left, right = 1, 1

    for i in range(n):
        results[i] *= left
        left *= nums[i]

    for i in range(n - 1, -1, -1):
        results[i] *= right
        right *= nums[i]
    return results


# Test
print(productExceptSelf([1, 2, 3, 4]))  # [24, 12, 8, 6]
print(productExceptSelf([2, 3, 4, 5]))  # [60, 40, 30, 24]
print(productExceptSelf([1, 0, 3, 4]))  # [0, 12, 0, 0]
